src/main/java/hudson/plugins/timestamper/pipeline/GlobalDecorator.java

```diff
@@ -0,0 +1,103 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.plugins.timestamper.pipeline;
+
+import hudson.Extension;
+import hudson.console.LineTransformationOutputStream;
+import hudson.model.Queue;
+import hudson.model.Run;
+import hudson.plugins.timestamper.TimestamperConfig;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.log.TaskListenerDecorator;
+
+/**
+ * Applies plain-text timestamp prefixes to all Pipeline log lines.
+ */
+public final class GlobalDecorator extends TaskListenerDecorator {
+    
+    private static final Logger LOGGER = Logger.getLogger(GlobalDecorator.class.getName());
+
+    // Almost ISO_OFFSET_DATE_TIME, but uses .SSS instead of .nnnnnnnnn to show milliseconds instead of nanoseconds and uses X instead of Z so the offset shows up as `Z` rather than `+0000`.
+    static final DateTimeFormatter UTC_MILLIS = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSX");
+
+    private static final long serialVersionUID = 1;
+
+    GlobalDecorator() {}
+
+    @Override
+    public OutputStream decorate(final OutputStream logger) throws IOException, InterruptedException {
+        return new LineTransformationOutputStream() {
+            @Override
+            protected void eol(byte[] b, int len) throws IOException {
+                synchronized (logger) { // typically this will be a PrintStream
+                    logger.write('[');
+                    logger.write(ZonedDateTime.now(ZoneOffset.UTC).format(UTC_MILLIS).getBytes(StandardCharsets.US_ASCII));
+                    logger.write(']');
+                    logger.write(' ');
+                    logger.write(b, 0, len);
```

Comment:
I noticed in local testing that all of these individual calls to `logger.write` can cause a high amount of interpolation for parallel branches in a Pipeline build. Maybe better to consolidate things into a single write call. I think the reason the synchronization didn't help is that `FileLogStorage` returns a new instance of a logger for each call to `FileLogStorage#overallListener` and `FileLogStorage#nodeListener`, but I'm not sure.

Here is an example of timestamp interleaving, which I noticed while looking at jenkinsci/workflow-job-plugin#111. Note the overlapping branch names on the first line:

<img width="521" alt="screen shot 2018-12-13 at 10 41 54" src="https://user-images.githubusercontent.com/1068968/49950517-f0658e80-fec5-11e8-90b4-b8d968c99b52.png">


Here's what I had in mind; it could probably be made more efficient using `ByteBuffer#array` where available or by copying into a `byte[]` directly without an intermediate `ByteBuffer`:
```
ByteBuffer buffer = ByteBuffer.allocate(1 + 26 + 1 + 1 + len);
buffer.put((byte)'[');
buffer.put(ZonedDateTime.now(ZoneOffset.UTC).format(UTC_MILLIS).getBytes(StandardCharsets.US_ASCII));
buffer.put((byte)']');
buffer.put((byte)' ');
buffer.put(b, 0, len);
byte[] array = new byte[buffer.position()];
buffer.flip();
buffer.get(array);
logger.write(array);
```

Reply:
Also note that this only really seems to be a problem if at least one of the parallel branches is running on the master, probably thanks to the remote log buffering in workflow-api.

Reply:
See jglick/timestamper-plugin#1.